[쇼핑 주문 3, 4단계] 환노 미션 제출합니다.

hwannow opened the pull request:
안녕하세요 둘리 !
3, 4단계 미션도 잘 부탁 드립니다 🙇‍♀️


📌 리뷰를 진행하시기에 앞서
---
1, 2단계와 동일하게 아래의 텍스트를 local.properties에 추가해야 합니다 💪

```kotlin
name="hwannow"
password="password"
```

# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->


https://github.com/user-attachments/assets/b669095d-bd2a-497d-b912-5c4436058ff4

hwannow committed:
refactor: 상품 아이템 업데이트 시 아이디로 비교하도록 변경

hwannow committed:
fix: 잘못된 Fragment ID 참조를 container ID로 수정

hwannow committed:
fix: 장바구니 담기 성공 시 이벤트 발생시키도록 수정

hwannow committed:
refactor: first를 firstOrNull로 변경하여 안전성 강화

hwannow committed:
refactor: API service에서 accept 헤더 제거

hwannow committed:
refactor: selectedIds에 방어적 복사 적용

hwannow committed:
refactor: var를 val로 변경

hwannow committed:
test: FakeCartProductRepository의 장바구니에 상품 추가 시 success 호출하도록 수정

hwannow committed:
refactor: action bar title에 string 리소스 사용

hwannow committed:
refactor: IllegalArgumentException에 에러 메시지 추가

hwannow committed:
fix: 최근 본 상품 목록 조회 실패 시 에러 핸들링하도록 변경

hwannow committed:
refactor: updateQuantity에 바꾸고 싶은 quantity를 받도록 변경

hwannow committed:
feat: 선택한 상품들을 한 번에 삭제하는 기능 추가

hwannow committed:
test: `deleteAll` 메서드 구현

hwannow committed:
build: 코루틴 관련 의존성 추가

hwannow committed:
feat: CartProductApiService에 코루틴 적용

hwannow committed:
test: repository 및 viewmodel 테스트에 coroutine test 적용

hwannow committed:
feat: ProductApiService에 코루틴 적용

hwannow committed:
test: repository 및 viewmodel 테스트에 coroutine test 적용

hwannow committed:
refactor: CartProductRemoteDataSource에서 콜백 방식 삭제

hwannow committed:
refactor: 비동기 처리 및 코드 개선

hwannow committed:
refactor: FakeRepository의 메서드 네이밍 변경

hwannow committed:
refactor: ShoppingCartFragmentFactory 삭제

hwannow committed:
refactor: repository의 suspend 함수를 IO 스레드에서 실행하도록 수정

hwannow committed:
refactor: 도메인 모델과 데이터 엔티티, DTO 간 변환 로직 위치 변경

hwannow committed:
fix: 장바구니 선택 페이지 이동 시 가격 및 개수 초기화 버그 수정

hwannow committed:
refactor: postValue를 setValue로 수정

hwannow committed:
refactor: isSelectedAll 상태를 products에 따라 계산하도록 변경정

hwannow committed:
feat: 선택된 상품이 있을 때만 주문 버튼 활성화

hwannow committed:
refactor: _selectedProduct -> _onSelectedProduct 네이밍 변경

hwannow committed:
refactor: 장바구니 추천 상품 수량 변경 로직 개선

hwannow committed:
refactor: 장바구니 수량 변경 및 선택 로직 개선

hwannow committed:
test: 불필요한 runTest 제거

hwannow committed:
refactor: `finishOrder` 함수를 viewModelScope로 변경

hwannow committed:
docs: 구현할 기능 목록 작성

hwannow committed:
refactor: 장바구니 추천 화면에서 상품 정보 전달 방식 변경

hwannow committed:
test: CartProductViewModelTest의 selectedIds를 selectedProducts로 변경

hwannow committed:
fix: selectedProduct가 변경될 때 totalPrice, totalCount가 변경되도록 수정

hwannow committed:
feat: 결제 화면 UI 구현

hwannow committed:
feat: 쿠폰 UI 구현

hwannow committed:
feat: 결제 화면 초기 설정

- PaymentActivity DataBinding 설정
- PaymentActivity `newIntent` 추가

hwannow committed:
feat: 주문하기 버튼 클릭 시 결제 화면으로 이동

hwannow committed:
feat: 결제 화면 ViewModel 설정 및 총 금액 표시

hwannow committed:
feat: 쿠폰 도메인 모델 정의

hwannow committed:
feat: 쿠폰 데이터 응답 DTO 클래스 구현

hwannow committed:
feat: 쿠폰 데이터 가져오기 기능 구현

hwannow committed:
feat: 결제 화면에서 쿠폰 목록을 가져오도록 구현

hwannow committed:
feat: 쿠폰 목록 표시 기능 추가

hwannow committed:
feat: 쿠폰 선택 기능 구현

hwannow committed:
feat: 유효한 쿠폰만 노출되도록 필터링

hwannow committed:
feat: 결제 상세 정보 표시 구현

hwannow committed:
feat: 쿠폰 할인 적용 로직 구현

hwannow committed:
feat: 주문하기 기능 구현

hwannow committed:
refactor: 사용하지 않는 메서드 및 변수 제거

hwannow committed:
docs: README 업데이트

hwannow committed:
feat: PaymentActivity에 Toolbar 뒤로가기 버튼 추가 및 제목 설정

hwannow committed:
test: 쿠폰 도메인 테스트 코드 작성

hwannow committed:
test: PaymentDetail에서 쿠폰 종류에 따른 할인 금액 계산 기능 테스트 작성

hwannow committed:
test: 결제 화면 뷰모델 테스트 코드 작성

hwannow commented on app/src/main/java/woowacourse/shopping/data/datasource/remote/CartProductRemoteDataSource.kt:
```diff
@@ -1,164 +1,85 @@
-                            onResult(Result.success(PagedResult(products, hasNext)))
-                        } ?: onResult(Result.success(PagedResult(emptyList(), false)))
-                    } else {
-                        onResult(Result.failure(Exception("HTTP ${response.code()}: ${response.message()}")))
-                    }
-                }
+    ): Result<PagedResult<CartProductDto>> {
```
[#108](https://github.com/woowacourse/android-shopping-order/pull/108#discussion_r2123042690)

저도 동의합니다 !

말씀하신 대로 `data` 계층에서 `domain model`에 의존하게 된다면
`domain`이 바뀌는 경우 `data` 계층도 바뀌게 되겠군요 🤔

`domain`과 `data`의 중간 다리인 `repository`에서 `dto -> domain `변환을 해 주도록 변경해 주었습니다 !

hwannow commented on app/src/main/java/woowacourse/shopping/view/cart/selection/CartProductSelectionFragment.kt:
[#108](https://github.com/woowacourse/android-shopping-order/pull/108#discussion_r2117615073)

해당 `fragment`의 `onStart`에서 데이터를 다시 로드하는 이유는
**추천 화면에서 뒤로가기 -> 선택 화면** 시 장바구니 상품이 재로드되어야 하기 때문입니다 !

사용자가 추천 화면에서 상품을 담고 돌아왔을 때 선택 화면에서 최신 상태가 반영되도록 하려고 했습니다

hwannow commented:
[#108 ](https://github.com/woowacourse/android-shopping-order/pull/108#discussion_r2123201130)
> dp와 sp의 차이가 무엇일까요?

`text`의 크기에는 일반적으로 `sp`를 사용합니다 !

하지만 이 코드에서 사용하는 `TextView`는 숫자만 표시되는 UI 요소이고,
디자인적으로 고정된 크기의 원 안에 정확한 사이즈를 구현해야 한다고 생각했습니다 🤔

따라서 사용자의 텍스트 크기 설정에 따라 크기가 변하지 않도록 `dp`를 사용하였습니다

--- 
[#108 ](https://github.com/woowacourse/android-shopping-order/pull/108#issuecomment-2935716721)

> 추가로 reflection이 왜 성능상 문제가 되는지, 다른 이점은 없는지도 한번 찾아보시죠 👀

`reflection`의 경우 동적으로 타입을 해석하기 때문에 JVM 최적화가 불가능하다고 합니다
( ex. 어떤 클래스의 메서드를 실행하고 싶은 경우 -> 클래스 이름으로 클래스 찾기 -> 메서드 이름으로 메서드 탐색 -> 메서드 호출 가능 여부 탐색 -> 실행 )

따라서` reflection`을 이용한 연산은 일반적인 코드보다 성능이 느리다고 하네요 😵‍💫

하지만 런타임에 클래스를 동적으로 다룰 수 있다거나 private 멤버에 접근하는 등,
유연성과 접근성 측면에서는 이점이 있다고 하네요 😵‍💫

hwannow committed:
refactor: xml에 함수를 넘기고, onClick 속성에서 함수를 사용하도록 변경

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/datasource/remote/OrderRemoteDataSource.kt:
```diff
@@ -0,0 +1,22 @@
+    suspend fun submitOrder(cartItemIds: List<Int>): Result<Unit> {
+        val response = orderApiService.submitOrder(body = OrderRequestDto(cartItemIds))
+
+        return if (response.code() == SUCCESS_POST) {
```
`HttpsURLConnection.HTTP_CREATED`와 같은 상수들도 활용해볼 수 있어요.

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/63e5c388f976ea5869ba7786383c0a7ab030296f)

이미 정의된 상수가 있었군요 🫢

`companion object`에 정의한 상수들을 모두 `HttpURLConnection`의 상수로 대체해 주었습니다 !

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/datasource/remote/ProductRemoteDataSource.kt:
```diff
@@ -1,105 +1,64 @@
+                return Result.success(body)
+            } else {
+                Result.success(null)
+            }
```
이 리뷰를 참고해주세요!

hwannow replied:
[refactor: 중복된 return 키워드 제거](https://github.com/woowacourse/android-shopping-order/pull/130/commits/58a09c7c147a00fe81b666daf9d428a39e4798d2)

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/repository/CartProductRepositoryImpl.kt:
```diff
@@ -1,69 +1,64 @@
+    ): Result<Int> =
+        withContext(Dispatchers.IO) {
+            remoteDataSource.insert(productId, quantity)
+        }
```
Dispatchers.IO로 디스패쳐를 바꿔주어야만 할까요? 🤔  (이하동일)
Retrofit suspend API에 대해 찾아보시죠!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/c96a2b8932df45c9f53f6412175bb61c9a0fcc5a)

`interface`로 작성한 `api` 메서드들은 `Retrofit2` 내부 코드에서 IO를 알아서 처리하고, 리턴한다고 하네요 🤔
( 알아서 `thread` 생성 -> 작업 )

하지만 `Room`을 이용한 작업을 하거나, 처리해야 할 데이터의 양이 많을 경우,
IO를 명시해 주어야 하는 경우도 있다고 합니다

hyemdooly replied:
맞아요, Retrofit이 알아서 해주는거라서 다른 작업을 할 때는 말씀하신대로 IO로 바꿔주어야할 수도 있어요~!

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/service/CartProductApiService.kt:
suspend 함수에서 body만을 리턴할 수도 있어요!
왜 그런지 한번 찾아보시죠 😄 

추가로 DataSource 코드에서 Result로 변환하는 코드가 반복되고 있는데 CallAdapter를 활용해보시는 것도 좋을 것 같아요!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/ef4a332eaddbdf424986e8ba9534482521c065ca)

<img width="741" alt="스크린샷 2025-06-09 01 38 30" src="https://github.com/user-attachments/assets/dbe209c0-97ac-46b2-b248-a51471c1bd4d" />

찾아 보니 `suspend` 함수를 지원하고, `body`를 직접 반환할 수도 있다고 하네요 😵‍💫

이전에 `Response`로 받은 건 응답 코드에 따른 예외를 발생시키기 위함이었는데,
`body`를 직접 반환하면 알아서 `HttpException`이 발생하는군요 ... !

시간이 부족해서 완전히 이해하고 쓴 코드는 아닌 것 같지만 🥲
`callAdapter`를 사용하여 `Result<T>`를 직접 반환하도록 해 주었습니다 !

hyemdooly replied:
맞아요 그래서 Result로 굳이 안해도, runCatching을 활용해서 예외처리를 할 수도 있어요~!

hyemdooly replied:
맞아요 그래서 Result로 굳이 안해도, runCatching을 활용해서 예외처리를 할 수도 있어요.
CallAdapter를 활용하면 개발자가 Exception을 정의해서, CallAdapter에서 변환해준 후 받은 쪽에서 어떤 Exception인지에 따라 특정 액션을 취해줄 수도 있구요!

hyemdooly commented on app/src/main/java/woowacourse/shopping/domain/model/PaymentDetail.kt:
```diff
@@ -0,0 +1,38 @@
+                is PercentageCoupon -> coupon.calculateDiscountAmount(orderAmount)
+                is BuyXGetYCoupon -> coupon.calculateDiscountAmount(selectedProducts)
+                is FreeShippingCoupon -> 0
+            }
```
쿠폰이 무조건 할인액수를 계산하는 책임을 가지고 있다면 부모클래스에 정의하고, 자식클래스에서 구현하면 어떨까요?

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/121d2cd68ffab614d014101d4fbbfa58f8d6900d)

기존에 부모 클래스에 정의해 주지 않았던 이유는 
각 쿠폰이 할인 금액을 계산하기 위해 필요로 하는 인자가 다르기 때문입니다 ! 😵‍💫

```kotlin
BuyXGetY -> 전체 아이템 리스트 필요 (List<CartProduct>)
Percentage -> 총 금액 (Int)
Fixed -> 필요없음
FreeShipping -> 필요없음
```

하지만 부모 클래스에 정의하고 자식 클래스에서 구현하니
이를 사용하는 `PaymentDetail` 입장에서는 코드가 깔끔해지네요 🤔

hyemdooly replied:
저도 인자가 달라서 따로 구현해주셨구나 생각했어요.ㅎㅎ
특정 케이스에서 굳이 필요없는 인자라도 액션이 공통됨 + 인자가 '보통' 필요하다면 저는 부모클래스에 정의해주는 편이긴 합니다!

hyemdooly commented on app/src/main/java/woowacourse/shopping/domain/model/coupon/BuyXGetYCoupon.kt:
```diff
@@ -0,0 +1,25 @@
+        val mostExpensiveItem = filteredItems.maxByOrNull { it.product.price } ?: return 0
+        return -1 * mostExpensiveItem.product.price * getQuantity
+    }
+}
```
할인 금액이 음수여야할지, 양수여야할지 고민해볼만한 부분인 것 같아요. 🤔

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/a2e71bc1106a524c33ab4eb4adcb095a215e6527)

다시 생각해 보니 음수로 저장하면 다른 개발자들이 보았을 때 혼란스러울 수 있을 것 같네요 😅

할인 금액을 양수로 저장하고, xm에서 음수로 나타내도록 수정해 주었습니다 !

hyemdooly commented on app/src/main/java/woowacourse/shopping/domain/model/CartProduct.kt:
```diff
@@ -1,10 +1,12 @@
     val product: Product,
     val quantity: Int = MINIMUM_PRODUCT_QUANTITY,
-) {
+) : Serializable {
```
Serializable을 사용한 이유가 궁금합니다!

hwannow replied:
`Selection -> Recommendation -> Payment` 간의 화면 전환에서
`CartProduct 리스트`를 `Intent & Bundle`을 통해 전달해야 하기 때문에 `Serializable`을 적용하였습니다 😵‍💫

hyemdooly replied:
아하 `Parcelable`도 있는데 `Serializable`를 선택한 이유가 궁금했어요 ㅎㅎ

hwannow replied:
도메인 모델에서 `Parcelable`를 사용하면 안드로이드 의존성이 생기기 때문에
`Parcelable`을 사용하지 않고 `Serializable`을 사용했습니다 !

도메인 레이어는 독립적이어야 한다고 생각합니다 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/recommendation/CartProductRecommendationViewModel.kt:
```diff
@@ -25,19 +29,25 @@ class CartProductRecommendationViewModel(
+    val onSelectedProduct: SingleLiveData<Product> get() = _onSelectedProduct
+
+    private val _onStartOrder = MutableSingleLiveData<List<CartProduct>>()
+    val onStartOrder: SingleLiveData<List<CartProduct>> get() = _onStartOrder
```
네이밍이 `onXXX`가 된 이유가 있을까요?
SingleLiveData로 만들어준 이유도 궁금해요. Event가 아닌 것 같아서요 🤔 
이벤트를 의미하는 것이라면 사용하는 쪽에서도 잘 파악할 수 있게, Event를 만들어 사용하면 어떨까요?

hwannow replied:
[refactor: startOrder를 SingleLiveData -> LiveData로 변경](https://github.com/woowacourse/android-shopping-order/pull/130/commits/8058627bd02498ebb4afb30fabfb1193f66c01e1)

`SingleLiveData`로 있을 필요는 없을 것 같군요 🤔

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/selection/CartProductSelectionViewModel.kt:
```diff
@@ -62,41 +67,98 @@ class CartProductSelectionViewModel(
-            loadPage(_page.value ?: FIRST_PAGE_NUMBER)
+
+        viewModelScope.launch {
+            val newQuantity = cartProductItem.cartProduct.quantity - QUANTITY_TO_ADD
```
여기서는 ADD는 아니라서, 상수 이름을 바꿔주면 어떨까요?

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/d64011fa8236bd4e9cef7d4f72a96cc7e0151658)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/selection/CartProductSelectionViewModel.kt:
```diff
@@ -105,58 +167,58 @@ class CartProductSelectionViewModel(
-        _isSinglePage.postValue(!hasNext && !hasPrevious)
+        _hasPrevious.value = page > FIRST_PAGE_NUMBER
+        _hasNext.value = hasNext
+        _isSinglePage.value = !hasNext && !hasPrevious
```
사실 여기 LiveData들도 page에 의존하고 있으니, `map`을 활용해볼 수 있을 것 같아요!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/48bcfdec6c188684f2622977b66c04a65eadf739)

`_hasPrevious, _isSinglePage`를 `map` 방식으로 변경해 주었습니다 !

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/payment/PaymentViewModel.kt:
```diff
@@ -0,0 +1,96 @@
+                Log.e("error", it.message.toString())
+            }
+        }
+    }
```
`SingleLiveData`를 활용해서 이벤트를 관리하는 변수를 만들고, 특정 이벤트가 발생할 때 토스트를 띄운다든지, 유저에게 정보를 알릴 수 있습니다.
활용해볼 부분이 있다면 활용해봐도 좋을 것 같아요!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/628dba5c7f81ca0a2c484e41fec71c8181242188)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/payment/PaymentViewModel.kt:
```diff
@@ -0,0 +1,96 @@
+
+            _onFinishOrder.setValue(Unit)
+        }
+    }
```
이런 것도 처리해보면 좋을 것 같아요.
만약 주문할 때 인터넷이 끊겨서 주문되지 않았다면?!

<img src="https://github.com/user-attachments/assets/00f693ce-9007-486b-bfc3-e0396d7123c4" width=300 />

실제로 쿠폰 선택 창에서 인터넷을 끊고 주문하기를 누르니 앱이 터져버리네요!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/531b6a79a589ea55acf7b7d9e8c9ebb00ace4bad)

`onSuccess` 내부에서 완료하도록 변경해 주었습니다 !

hyemdooly commented on app/src/test/java/woowacourse/shopping/fixture/FakeCartProductRepository.kt:
```diff
@@ -21,56 +20,46 @@ class FakeCartProductRepository : CartProductRepository {
-            newQuantity == 0 -> delete(cartProduct.id) { onResult(Result.success(Unit)) }
+        newQuantity: Int,
+    ): Result<Unit> {
+        return when {
+            newQuantity == 0 -> return delete(cartProduct.id)
```
요것도 참고해주세요~!

hwannow replied:
[refactor: 중복된 return 키워드 제거](https://github.com/woowacourse/android-shopping-order/pull/130/commits/58a09c7c147a00fe81b666daf9d428a39e4798d2)

hyemdooly reviewed (changes requested):
환노 안녕하세요~! 빠르게 3, 4단계 구현을 해주셨군요 👍 
전체적으로 잘 작성된 것 같아요. 테스트 코드도 다 작성해주신 것 같네요! 💯 
추가로 고민해보면 좋을 부분들에 대해 코멘트 남겨드렸습니다.
큼직한 내용들이 될 것 같은데, 이번 기회에 공부하면 큰 도움이 될 내용들이니 한번 시도해보세요!
레벨2 마지막까지 화이팅입니다! 💪

hwannow committed:
refactor: Coupon 인터페이스에 `calculateDiscountAmount` 메서드 추가

hwannow committed:
refactor: 할인 금액 저장 방식 수정

- 할인 금액을 음수에서 양수로 변경
- UI에 표시될 때 -1을 곱하여 할인 금액을 음수로 표시

hwannow committed:
refactor: 상수명 변경 `QUANTITY_TO_ADD` -> `QUANTITY_STEP`

hwannow committed:
refactor: hasPrevious, isSinglePage를 map 방식으로 변경

hwannow committed:
refactor: HTTP 상태 코드를 HttpURLConnection 상수로 대체

hwannow committed:
refactor: 중복된 return 키워드 제거

hwannow committed:
refactor: CallAdapter 적용

hwannow committed:
test: `hasPrevious`, `hasNext`, `products` LiveData observe 추가

hwannow committed:
feat: 에러 발생 시 토스트 메시지 표시

hwannow committed:
refactor: 주문 완료 이벤트 발생 시점 변경

hwannow committed:
refactor: startOrder를 SingleLiveData -> LiveData로 변경

hwannow committed:
refactor: Repository에 withContext 삭제

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/network/ResponseAdapterFactory.kt:
```diff
@@ -0,0 +1,26 @@
+        retrofit: Retrofit,
+    ): CallAdapter<*, *>? {
+        if (Call::class.java != getRawType(returnType)) return null
+        check(returnType is ParameterizedType)
```
에러 메시지도 같이 포함하면 좋을 것 같아요!

hwannow replied:
[refactor: ResponseAdapterFactory에 에러 메시지 추가](https://github.com/woowacourse/android-shopping-order/pull/130/commits/a7d397ead1daee108f4d08d1cafa0c35d532e03b)

hyemdooly commented on app/src/main/java/woowacourse/shopping/data/repository/RecentProductRepositoryImpl.kt:
```diff
@@ -1,70 +1,69 @@
+                val productIds = entities.map { it.productId }
+
+                val productResult = productRepository.getProductsByIds(productIds)
+                val products = productResult.getOrElse { throw it } ?: emptyList()
```
`getOrThrow()`를 활용해볼 수 있어요!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/c8764ecc8bb999f893276aff17ee2e0b225c6233)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/recommendation/CartProductRecommendationViewModel.kt:
```diff
@@ -25,150 +30,201 @@ class CartProductRecommendationViewModel(
+    private val _startOrder = MutableLiveData<List<CartProduct>>()
+    val startOrder: LiveData<List<CartProduct>> get() = _startOrder
+
+    private val _onSelectedProduct = MutableSingleLiveData<Product>()
+    val onSelectedProduct: SingleLiveData<Product> get() = _onSelectedProduct
+
+    private val _onError = MutableSingleLiveData<Error>()
+    val onError: SingleLiveData<Error> get() = _onError
```
보통 `onXXX`은 callback 함수 이름에 많이 사용합니다!

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/043b3668bfd8958d5bfb7993a65adc9aa8aff523)

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/cart/recommendation/CartProductRecommendationViewModel.kt:
```diff
@@ -25,150 +30,201 @@ class CartProductRecommendationViewModel(
-                                        .map { ProductItem(it) }
-                                        .toList()
-
-                                _recommendedProducts.postValue(recommended)
-                            }.onFailure {
-                                Log.e("error", it.message.toString())
-                            }
-                    }
+                    recentProduct ?: return@launch
+                    val pagedProductsResult = productRepository.getPagedProducts()
+
```
이렇게 되면 `getLastViewedProduct()`와 `getPagedProducts()`가 순차적으로 실행될 것 같은데요.
async를 활용해보면 어떨까요?

hwannow replied:
[해당 내용 관련 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/130/commits/c8764ecc8bb999f893276aff17ee2e0b225c6233)

hyemdooly replied:
커밋이 잘못 첨부된 것 같아요!ㅎㅎ

hyemdooly commented on app/src/main/java/woowacourse/shopping/view/util/Error.kt:
에러 정의 👍 
나중에는 이 Error도 무궁무진하게 늘어날 수 있어서 화면 or 관심사별로 분리해두는 것도 방법이겠네요!
에러 뿐만 아니라 성공했을 때라든지, 다른 이벤트를 처리할 수도 있겠어용

hyemdooly commented:
[로드실패.webm](https://github.com/user-attachments/assets/e1e69e96-6923-43d5-8a6e-d7fc4d478c6e)

[Screen_recording_20250610_133614.webm](https://github.com/user-attachments/assets/4788e012-7db1-46be-9b42-94552842e454)

에러 토스트가 자꾸 뜨는 버그가 있어요!

hyemdooly reviewed (changes requested):
환노 안녕하세요! 빠르게 리뷰 반영을 해주셨네요 👍 
추가로 버그 제보😱와 고민해볼만한 부분들에 대해 리뷰 드렸습니다.
마지막까지 얼마남지 않았습니다! 화이팅이에요 💪

hwannow committed:
refactor: ResponseAdapterFactory에 에러 메시지 추가

hwannow committed:
refactor: getOrElse 대신 getOrThrow 사용

hwannow committed:
refactor: 최근 본 상품과 추천 상품을 병렬로 로드하도록 수정

hwannow committed:
refactor: LiveData 변수명 변경

_onError -> _errorEvent
_onFinishOrder -> _orderComplete
_onSelectedProduct -> _selectedProduct

hyemdooly reviewed (approved):
환노 안녕하세요! 리뷰 반영 빠르게 해주셨네요 👍  이번 미션은 여기서 마무리하겠습니다.
이번 미션을 통해서 많은 것을 얻어가셨기를 바랍니다. 🥹

레벨2 마지막 미션도 고생 많으셨습니다! 우테코 생활 중 궁금한 점이나 의견 나누고 싶은 내용이 있다면 언제든지 DM 주세요~
앞으로의 우테코 여정도 응원하겠습니다! 💪
